src/ucs/datastruct/lru.h

```diff
@@ -0,0 +1,132 @@
+/**
+* Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2001-2023. ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifndef __LRU_H__
+#define __LRU_H__
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <ucs/type/status.h>
+#include <ucs/datastruct/array.inl>
+#include <ucs/datastruct/khash.h>
+#include <ucs/debug/assert.h>
+
+
+/* LRU element data structure */
+typedef struct {
+    /* Key to use as hash table input */
+    const void     *key;
+    /* Linked list item */
+    ucs_list_link_t list;
+} ucs_lru_element_t;
+
+
+KHASH_INIT(lru_hash, uint64_t, ucs_lru_element_t, 1, kh_int64_hash_func,
+           kh_int64_hash_equal)
+
+/* Hash table type for LRU cache */
+typedef khash_t(lru_hash) ucs_lru_hash_t;
+
+/* LRU cache data structure */
+typedef struct ucs_lru {
+    /* Hash table of addresses as keys */
+    ucs_lru_hash_t  hash;
+    /* Linked list ordered by recent access */
+    ucs_list_link_t list;
+    /* Cache size */
+    size_t          size;
+    /* Number of elements currently in cache */
+    size_t          capacity;
+} ucs_lru_t;
+
+typedef struct ucs_lru *ucs_lru_h;
+
+UCS_ARRAY_DEFINE_INLINE(lru, unsigned, void*);
+
+/**
+ * @brief Create a new LRU cache instance.
+ *
+ * @param [in]  capacity  Max number of elements to store.
+ * @param [out] lru_p     Pointer to the allocated lru struct.
+ *
+ * @return UCS_OK if successful, or an error code as defined by
+ * @ref ucs_status_t otherwise.
+ */
+ucs_status_t ucs_lru_create(size_t capacity, ucs_lru_h *lru_p);
+
+/**
+ * @brief Destroy LRU cache.
+ *
+ * @param [in] lru  Handle to the LRU cache.
+ */
+void ucs_lru_destroy(ucs_lru_h lru);
+
+static UCS_F_ALWAYS_INLINE void ucs_lru_pop(ucs_lru_h lru)

```

Comment:
It is static, inlined and used only once.
Maybe remove it & replace w inlining the code?
A justification for a "no" could be if you think it might be useful in the future as a stand-alone func.

Reply:
BTW as a bonus, then can rename `_remove` to `_pop` which would be even more correct naming.